### PR TITLE
chore: Tox configuration scripts

### PR DESCRIPTION
**Pull request by tinvaan**
## What kind of change does this PR introduce?

Project maintenance, dev workflow enhancement.

## What is the current behavior?

https://github.com/supabase-community/supabase-py/issues/696

## What is the new behavior?

Move away from `Makefiles` to `tox` configuration scripts. `tox` is tailor made for python projects compared to `Makefiles` and allows for venv management and test automation + standardization in CI workflows.

For eg, the below config will run our tests against multiple python environments, to help guard against any breaking changes we introduce and thus ensuring backward compatibility.
```ini
[tox]
isolated_build = True
envlist = format, py37, py38, py39, py310, py311, py312
```

## Additional context

Tests against all supported versions can be launched using a single `tox` command and optionally running tests against specific versions is also supported using the `tox -e` argument.

### CODE REVIEW

**Comment by tinvaan** [comment]
Sorry for the noise, CI finally looks green ✅

File: tox.ini [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+poetry_dep_groups =
+    dev
+setenv =
+    PYTHONPATH = {toxinidir}
+passenv =
+    SUPABASE_URL
+    SUPABASE_KEY
+commands =
+    poetry install --no-root -v
+    poetry run pytest -s --disable-warnings --cov=./ --cov-report=term --cov-report=html
```
[the remark this answers is not in the document]
Comment (tinvaan) [reply tox.ini:24]: Yeah, I had left it in place so as to not pollute the CI logs. Good spot and thanks for flagging.

**Comment by tinvaan** [comment]
@mansueli , @J0 -- any further thoughts on this?

**Comment by J0** [comment]
Hey @tinvaan,

Thanks for putting this together - it's a little late where I'm at but I'll have a deeper look soon and get back

File: Makefile [hunk 2]
```diff
@@ -1,19 +0,0 @@
...
-
-tests_pre_commit:
-	poetry run pre-commit run --all-files
-
-run_tests: tests
-
-tests_only:
-	poetry run pytest --cov=./ --cov-report=xml --cov-report=html -vv
-
-build_sync:
```
Comment (J0) [inline Makefile:18]: Will we be adding an equivalent `tox` command for `build_sync`? Or is there a new proposed flow there

Comment (tinvaan) [reply Makefile:18]: aah good point. There is a new proposed flow that I have in mind is to unify the sync and async clients: https://github.com/supabase-community/supabase-py/issues/697

But that's quite a complex change that will need a fair bit of testing + review before it's release ready. For now, it definitely makes sense to keep a corresponding `tox` command for compatibility 👍🏼

Comment (tinvaan) [reply Makefile:18]: hey @J0! sorry for the delay, life happened. I've addressed this now -- will mark as resolved.

**Comment by tinvaan** [comment]
Resolved merge conflicts & painted CI green again.

/cc @J0

**Comment by J0** [comment]
@tinvaan,

Thanks for the PR!  I think the maintaining team will need to look through this before we can move forward as it will change the development flow  It may take a while.

cc: @silentworks @mansueli

**Comment by silentworks** [comment]
Closing this PR for the reasons given on https://github.com/supabase/supabase-py/issues/696#issuecomment-2306620767.

**Pull request closed without merging**